sdk/sdk.ts

```diff
@@ -0,0 +1,171 @@
+import { isPostMessageEvent, PostMessageEventName } from '../src/models/sdk'
+import { QueryParamKeys } from '../src/models/shell'
+import { isColorMode, UserSettings } from '../src/models/settings'
+import { iframeFeatureAllowList } from '../src/config/iframeFeatureAllowList'
+
+export const defaultRoot = 'https://chitchatter.im/'
+
+// NOTE: This is a subset of standard iframe attributes:
+// https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#attributes
+const iframeAttributes = [
+  'height',
+  'referrerpolicy',
+  'sandbox',
+  'style',
+  'width',
+]
+
+enum ChatEmbedAttributes {
+  COLOR_MODE = 'color-mode',
+  PLAY_MESSAGE_SOUND = 'play-message-sound',
+  ROOM_NAME = 'room',
+  ROOT_URL = 'root-url',
+  USER_ID = 'user-id',
+  USER_NAME = 'user-name',
+}
+
+const pollInterval = 250
+const pollTimeout = 10_000
+
+class ChatEmbed extends HTMLElement {
+  private iframe = document.createElement('iframe')
+
+  static get observedAttributes() {
+    const chatAttributes = Object.values(ChatEmbedAttributes)
+    const iframeAttributes = ['height', 'width']
+
+    return [...chatAttributes, ...iframeAttributes]
+  }
+
+  get chatConfig() {
+    const chatConfig: Partial<UserSettings> = {}
+
+    if (this.hasAttribute(ChatEmbedAttributes.USER_ID)) {
+      chatConfig.userId = this.getAttribute(ChatEmbedAttributes.USER_ID) ?? ''
+    }
+
+    if (this.hasAttribute(ChatEmbedAttributes.USER_NAME)) {
+      chatConfig.customUsername =
+        this.getAttribute(ChatEmbedAttributes.USER_NAME) ?? ''
+    }
+
+    chatConfig.playSoundOnNewMessage = Boolean(
+      this.hasAttribute(ChatEmbedAttributes.PLAY_MESSAGE_SOUND)
+    )
+
+    const colorMode = this.getAttribute(ChatEmbedAttributes.COLOR_MODE) ?? ''
+
+    if (isColorMode(colorMode)) {
+      chatConfig.colorMode = colorMode
+    }
+
+    return chatConfig
+  }
+
+  get rootUrl() {
+    return this.getAttribute(ChatEmbedAttributes.ROOT_URL) ?? defaultRoot
+  }
+
+  private sendConfigToChat = () => {
+    const { iframe, rootUrl } = this
+    const { origin: rootUrlOrigin } = new URL(rootUrl)
+
+    iframe.contentWindow?.postMessage(
+      {
+        name: PostMessageEventName.CONFIG,
+        payload: this.chatConfig,
+      },
+      rootUrlOrigin
+    )
+  }
+
+  private async sendConfigToChatAndWaitForConfirmation() {
+    const { rootUrl } = this
+
+    let timer: NodeJS.Timer
+    const { origin: rootUrlOrigin } = new URL(rootUrl)
+
+    const handleMessageReceived = (event: MessageEvent) => {
+      if (rootUrlOrigin !== event.origin) return
+      if (!isPostMessageEvent(event)) return
+      if (event.data.name !== PostMessageEventName.CONFIG_RECEIVED) return
+
+      clearInterval(timer)
+      window.removeEventListener('message', handleMessageReceived)
+    }
+
+    window.addEventListener('message', handleMessageReceived)
```

Comment:
Since you're adding an event listener to the window, you may want to add a `disconnectedCallback` that removes the event listener and clears the timer if the component gets removed from the DOM before a message is received.

Otherwise, there's a possibility that the timer or the event handler will retain the component in memory, even if it has been removed from the DOM.

Reply:
Great catch. This is fixed in https://github.com/jeremyckahn/chitchatter/commit/75cfca69d8ebeb9fce20d29e6a51541489170436!